### PR TITLE
Improve `readme.md`

### LINKED ISSUE #1069: Suggested documentation improvements
A quick brain dump of all the questions that arose while getting started with `next.js`. Some of these might be answered already and I just haven't looked hard enough. 

## How `next.js` improves performance
As I see it a major responsibility of the `next.js` build process and server is performance improvements.
Some are documented and some not and there doesn't seem to be a central place to look up all that is done.

I'd like to see a comprehensive list (possibly with reasoning and resources) of all the steps the build process and server takes inherently (e.g. compression, HTTP headers, minification, code splitting, etc.). And also steps it doesn't take (yet), pitfalls and configuration/extensions that the end user might have to implement by themselves. The Wiki might be a good place for such a list.

## Configuration File
The `README.md`'s [custom configuration](https://github.com/zeit/next.js#custom-configuration) section seems incomplete in terms of describing what keys the exported object from `next.config.js` accepts. Based on present documentation there is only one key `webpack`. Is that all there is for the moment?

## Programmatic API
A start to document `next.js`' programmatic API is taken in section [Custom server and routing](https://github.com/zeit/next.js#custom-server-and-routing) of `README.md`. What I'm missing is proper documentation for instance methods like `getRequestHandler()`, `render()` etc.

---

I'm more than happy to help with improvements but would like some suggestions and pointers from somebody more familiar with the matter. Also maybe the community has more points to add to the list.

> Documentation is like sex: when it is good, it is very, very good; and when it is bad, it is better than nothing.

### LINKED ISSUE #1134: Document next/error
I don't think it's documented anywhere?

Also (unrelated to title, sorry), I guess it doesn't support any kind of message localization.

### PR DESCRIPTION
**Pull request by niksurff**
I tried to improve the docs under the following premises:

* Strong introduction
* Detailed Getting Started guide (especially for beginners)
* Improved consistency and clarity in language and appearance (layout, typography)
* Add missing docs, especially: #1069, #1134
* Link to resources in and outside the document as much as possible

What's missing:

* CLI docs (e.g. `next init`)
* ...?

Honestly, once started, I got carried away and might have gone a bit overboard with the changes.
But what's done is done and I might as well share what I've got.

As always, these are suggestions: Keep what you like, discard the rest. ❤️ 

ℹ️ [View the updated `readme.md` here.](https://github.com/nikvm/next.js/blob/fix/docs/readme.md)

Fixes #1134 
Fixes #1069

### CODE REVIEW

**Review by timneutkens — COMMENTED** [review]
I like the improvements. Only thing I notice is that the docs feel 'crowded' because the sub-sub headings are as small as the content. Could you have a look at this?

**Comment by timneutkens** [comment]
Found another missing docs ticket: https://github.com/zeit/next.js/issues/630#issuecomment-270040021

**Comment by sedubois** [comment]
Well if you want to add more docs there's also https://github.com/zeit/next.js/issues/1229 😄 but personally I'd consider the PR already good enough, then refine things in a new one.

**Comment by nikvm** [comment]
Thanks guys for having a read through.

@timneutkens I'm guessing you are especially looking at `h4` headings. Unfortunately this is Github's styling which we can't easily change. Using different heading levels breaks semantics. One change I was thinking about earlier is numbered sections and/or linked headings (appearing in blue). [Here is an example](https://github.com/jkbrzt/httpie/blob/master/README.rst). What do you think?

As for more missing docs: I'm with @sedubois and would rather get this PR approved and merged as is. And make changes in following PRs.

**Comment by sedubois** [comment]
I also think that the README is starting to show its limits. Ideally Next.js needs a proper documentation website... Then starting from there, it could later also have a nice gallery of examples, with screenshots and a link to each of their versions deployed on Now.

**Comment by timneutkens** [comment]
@sedubois I like this idea. 

@nikvm yeah I know 😞 The numbered sections thing is better indeed.

**Comment by nikvm** [comment]
@rauchg mentioned [here](https://github.com/zeit/next.js/issues/1069#issuecomment-279271542) that he'd like to stick with the readme. I don't mind either way. Personally I like [Gitbook](https://www.gitbook.com/). But these options always come with added maintenance overhead&mdash;keeping code and docs in one place is always nice. Plus `readme.md` comes bundled with the `npm` package (which is good for offline usage).

@timneutkens I'll look into how we might automate numbered sections as `tocdoc` (used to automate creating our T.O.C.) doesn't provide the functionality.

File: readme.md [hunk 1]
```diff
@@ -341,50 +341,64 @@ Router.onRouteChangeError = (err, url) => {
...
+      <li><Link href='/'><a>Home</a></Link></li>
+      <li><Link href='/about'><a>About</a></Link></li>
+      <li><Link href='/contact'><a>Contact</a></Link></li>
     </ul>
   </nav>
 )
 ```
 
-#### Imperatively
+When this higher-level `<Link>` component is first used, the `ServiceWorker` gets installed.
```
Comment (HaNdTriX) [inline readme.md]: I think this is obsolete since https://github.com/zeit/next.js/pull/957

Comment (arunoda) [inline readme.md]: Yep.

**Comment by bdefore** [comment]
This is great @nikvm, thanks!

If I could add one further request: this PR improves the warning somewhat but I think it should be more prominent and to link directly to the README of the latest stable release. I initially missed this when starting with next and tried using the beta functionality of `next.config.js` unawares and silently ignored.

File: readme.md (hunk 1 above)
Comment (nikvm) [inline readme.md]: I pushed [an update](https://github.com/zeit/next.js/pull/1256/commits/48b454605733c60b759038720cab230332b3e527) to address this.

**Comment by nikvm** [comment]
@bdefore How should we make it more obvious? Add more ⚠️ signs? Move it up?

@timneutkens I thought a little more a bout numbered sections. The problem I see is when adding sections not to the end of a level. This might break links referencing a particular issue. As headings get auto-linked it's also not trivial to exclude the numbering. So I still hesitate implementing numbering.

**Comment by sedubois** [comment]
@timneutkens how about merging this? It's going to get more and more conflicts otherwise

File: readme.md [hunk 2]
```diff
@@ -253,24 +278,34 @@ export default () => (
...
-- `replace(url, as=url)` - performs a `replaceState` call with the given url
+#### Using `url` Property
 
-The second `as` parameter for `push` and `replace` is an optional _decoration_ of the URL. Useful if you configured custom routes on the server.
+For routing imperatively each page component receives a `url` property, an object with the following API:
 
-#### Imperatively
+- `pathname: string` &ndash; Current path excluding the query string
+- `query: Object` &ndash; Parsed query string. Defaults to `{}`.
+- `push(url: string, as?: string): boolean` &ndash; Performs a [`pushState`](https://developer.mozilla.org/en-US/docs/Web/API/History_API#Example_of_pushState()_method) call with the given url. Returns `true` on success, else `false`.
```
Comment (rauchg) [inline readme.md]: we've deprecated `push` and `replace`, we want people to use `Router.push` and `Router.replace`

File: readme.md [hunk 3]
```diff
@@ -1,58 +1,80 @@
...
+npm init -y
+```
+
+Hereafter it takes __three simple steps__ to set up your app.
 
-### Setup
+### Install the `next` Package
 
-Install it:
+First up, install Next.js and it's peer dependencies via [npm](https://npmjs.com/package/next):
```
Comment (rauchg) [inline readme.md]: `its`

**Review by rauchg — CHANGES_REQUESTED** [review]
Two changes and we're good to go! Thank you so much, sorry for the delay

**Comment by timneutkens** [comment]
@nikvm please merge https://github.com/nikvm/next.js/pull/2 then I can merge this in. Thanks! 🤗

**Comment by bdefore** [comment]
@nikvm sorry i'm so late to respond, but at this point it's not important to highlight the difference between master and stable release. looking forward to the major bump!

**Comment by timneutkens** [comment]
@bdefore we will remove it right before 2.0 is released.

**Comment by timneutkens** [comment]
@nikvm yeah it includes those. I'm curious what conflicts it has now :(

**Comment by timneutkens** [comment]
Going to close this. We should port some of it at a later point.

**Pull request closed without merging**